File: Linked_Lists/7_2_reverse_single_sublist.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        dummy_head = sublist_head = ListNode(0, head)
        
        for _ in range(1, m):
            sublist_head = sublist_head.next
            
        sublist_iter = sublist_head.next
        
        for _ in range(n - m):
            temp = sublist_iter.next
            sublist_iter.next, temp.next, sublist_head.next = temp.next, sublist_head.next, temp
            
        return dummy_head.next

File: Linked_Lists/7_2_reverseSingleSublist.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    if(head == null){
        return null;
    }
    
    let prev = head;
    
    for(let i = 1; i <= m-1; i++){
// look for prev node
        prev = prev.next;
    }
    
    let curr = prev.next;
    let anchor = prev, tail = curr;
    
//     reverse sublist
    for(let i = 1; i < n-m; i++){
        let tempNode = curr.next;
// point backward
        curr.next = prev;
// move on
        prev.next = curr;
        curr = tempNode
    }
    
    if(anchor != null){
        anchor.next = prev;
    } else {
        head = prev;
    }
    
    tail.next = curr;
    return head;
};


/*
Sigle Pass:
- get the anchor node
- reverse the sublist with a dummy head
- attach back de sublist
1->2->3->4-> null
dummyHead: null<-2<-3<-

1->3....2->4

Time: O(n)
Space: O(1)

*/